README


This archive contains the distribution PowerDNS-API-Client,
version 0.03:

  Client for PowerDNS::API

This software is copyright (c) 2011 by Ask Bjørn Hansen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for PowerDNS-API-Client
 
0.03  February 16, 2011
      - Require namespace::clean 0.20
      - Remove debug

0.01  February 14, 2011
      - First release


lib/PowerDNS/API/Client.pm
package PowerDNS::API::Client;
BEGIN {
  $PowerDNS::API::Client::VERSION = '0.03';
}
use strict;
use warnings;

use warnings;
use strict;
use Moose;
use LWP::UserAgent ();
use URI ();

use namespace::clean -except => 'meta';

use PowerDNS::API::Client::Request;

has 'user' => (
    isa => 'Str',
    is  => 'ro',
    required => 1,
);

has 'password' => (
    isa => 'Str',
    is  => 'ro',
    required => 1,
);

has 'server' => (
    isa => 'Str',
    is  => 'rw',
);

has 'ua' => (
    is  => 'rw',
    isa => 'LWP::UserAgent',
    lazy_build => 1,
);

sub _build_ua {
    my $self = shift;
    my $ua = LWP::UserAgent->new();
    $ua->env_proxy;
    my $realm = URI->new($self->server)->host_port;
    $ua->credentials($realm, 'PowerDNS::API', $self->user, $self->password);
    return $ua;
}

sub _request {
    my ($self, $method, $path, %args) = @_;
    return PowerDNS::API::Client::Request->new(
        method => $method,
        path   => $path,
        args   => \%args,
        api    => $self,
    );

}

sub call {
    my $self = shift;
    my $http_response = $self->ua->request( $self->_request(@_)->http_request );
    return PowerDNS::API::Client::Response->new(http => $http_response)->data;
}


__PACKAGE__->meta->make_immutable;

local ($PowerDNS::API::Client::VERSION) = ('devel') unless defined $PowerDNS::API::Client::VERSION;

1;

__END__

=pod

=encoding utf8

=head1 NAME

PowerDNS::API::Client - Client for PowerDNS::API

=head1 SYNOPSIS

    my $client = PowerDNS::API::Client->new( server => 'https://api.example.com/' );

=head1 METHODS

=head2 call( $endpoint, %args )

Calls the endpoint (see the API documentation) with the specified
arguments.  Returns a hash data structure with the API results.

=head1 DEBUGGING


=head1 AUTHOR

Ask Bjørn Hansen, C<< <ask at develooper.com> >>

=head1 BUGS

Please report any bugs or feature requests to the issue tracker at
L<http://github.com/devel/PowerDNS-API-Client/issues>.


=head1 COPYRIGHT & LICENSE

Copyright 2011 Ask Bjørn Hansen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut



lib/PowerDNS/API/Client/Request.pm
package PowerDNS::API::Client::Request;
BEGIN {
  $PowerDNS::API::Client::Request::VERSION = '0.03';
}
use Moose;
use URI ();
# use Digest::SHA qw(hmac_sha256_hex);
use HTTP::Request ();
use namespace::clean -except => 'meta';

use PowerDNS::API::Client::Response;

has 'api' => (
    isa      => 'PowerDNS::API::Client',
    is       => 'ro',
    required => 1,
);

has method => (
    isa      => 'Str',
    required => 1,
    is       => 'ro',
);

has path => (
    isa      => 'Str',
    required => 1,
    is       => 'ro',
);

has args => (
    isa => 'HashRef[Str]',
    is  => 'rw',
);

sub _query {
    my %args = @_;

    #my $api_secret = delete $args{api_secret};
    #$args{api_ts} = time;
    #$args{api_sig} =
    #  hmac_sha256_hex(_get_parameter_string(\%args), $api_secret);

    my $uri = URI->new();
    $uri->query_form(%args);
    return $uri->query;
}

sub http_request {
    my $self = shift;
    my $uri  = URI->new($self->api->server);
    $uri->path("/api/" . $self->path);

    my $content = _query(
        %{$self->args},
        #api_key    => $self->api->api_key,
        #api_secret => $self->api->api_secret,
    );

    my $request = HTTP::Request->new($self->method => $uri);

    $request->header('Content-Type' => 'application/x-www-form-urlencoded');
    if (defined($content)) {
        $request->header('Content-Length' => length($content));
        $request->content($content);
        warn "PowerDNS::API::Client Request Content: $content\n"
          if $ENV{API_DEBUG};
    }

    return $request;
}

sub _get_parameter_string {
    my $args = shift;

    my $str = "";
    for my $key (sort { $a cmp $b } keys %{$args}) {
        next if $key eq 'api_sig';
        my $value = (defined($args->{$key})) ? $args->{$key} : "";
        $str .= $key . $value;
    }
    return $str;
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=encoding utf8

=head1 NAME

PowerDNS::API::Client::Request - Request object for PowerDNS::API::Client

=head1 SYNOPSIS

This class manages setting up requests for the PowerDNS::API::Client,
including signing of requests.

No user servicable parts inside.  This part of the API is subject to change.

    my $req = PowerDNS::API::Client::Request->new
       (api    => $powerdns_api,
        method => 'location/detail',
        args   => { foo => 'bar',
                    fob => 123,
                  },
       );

    my $http_request = $req->http_request;

=head1 METHODS

=head2 api

=head2 http_request

Returns a HTTP::Request version of the request.

=head1 AUTHOR

Ask Bjørn Hansen, C<< <ask at develooper.com> >>

=head1 COPYRIGHT & LICENSE

Copyright 2011 Ask Bjørn Hansen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut




lib/PowerDNS/API/Client/Response.pm
package PowerDNS::API::Client::Response;
BEGIN {
  $PowerDNS::API::Client::Response::VERSION = '0.03';
}
use Moose;
use JSON qw(decode_json);
use namespace::clean -except => 'meta';

has http => (
    is       => 'ro',
    isa      => 'HTTP::Response',
    required => 1,
);

has data => (
    is         => 'ro',
    isa        => 'HashRef',
    lazy_build => 1,
);

sub _build_data {
    my $self = shift;
    if ($self->http->content_type ne 'application/json') {
        if ($ENV{API_DEBUG}) {
            require Data::Dumper;
            warn "PowerDNS::API::Client Response: ", Data::Dumper::Dumper($self->http);
        }
        return +{
            http_status => $self->http->code,
            error       => $self->http->status_line,
        };
    }
    my $data = decode_json($self->http->decoded_content);
    $data->{http_status} = $self->http->code;
    if ($ENV{API_DEBUG}) {
        require Data::Dumper;
        warn "PowerDNS::API::Client Response: ", Data::Dumper::Dumper($data);
    }
    return $data;
}

__PACKAGE__->meta->make_immutable;

1;


__END__

=pod

=encoding utf8

=head1 NAME

PowerDNS::API::Client::Request - Request object for PowerDNS::API::Client

=head1 SYNOPSIS

This class manages setting up requests for the PowerDNS::API::Client,
including signing of requests.

No user servicable parts inside.  This part of the API is subject to change.

=head1 METHODS

=head2 api

=head2 http_request

Returns a HTTP::Request version of the request.

=head1 AUTHOR

Ask Bjørn Hansen, C<< <ask at develooper.com> >>

=head1 COPYRIGHT & LICENSE

Copyright 2011 Ask Bjørn Hansen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Client for PowerDNS::API',
  'AUTHOR' => 'Ask Bjørn Hansen <ask@develooper.com>',
  'BUILD_REQUIRES' => {
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'PowerDNS-API-Client',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'PowerDNS::API::Client',
  'PREREQ_PM' => {
    'LWP::UserAgent' => '0',
    'Moose' => '1.0',
    'URI' => '0',
    'namespace::clean' => '0.20'
  },
  'VERSION' => '0.03',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



